### ISSUE TITLE
Allow users to configure types for ContinueOnFailure

### ISSUE THREAD

**Issue by apoorvam** [getgauge/gauge-java#66]
Refer getgauge/gauge#499

Currently users can configure Steps such that they do not break on failure. This is done via an Annotation in Java.

Allow users to specify what type of errors should be honoured in such cases.

**Comment by apoorvam** [comment]
### Problem: 
The feature Continue on Failure will continue executing further steps if a step is marked with annotation `@ContinueOnFailure`. But with this, it cannot differentiate assertion errors and any other type of errors(say RuntimeError). 

### Solution: 
* Have a configurable property `gauge_cof_exception_whitelist`(Better property name?) in `java.properties` file as:
```
# specify the exceptions to be whitelisted for continue on failure
# you can specify multiple exceptions separated with a comma (,)
gauge_cof_exception_whitelist = java.lang.AssertionError
```
* `java.lang.AssertionError` will be set by default
* Gauge-Java looks for the list of exceptions in the above property
* If the current exception is one in the list and the current step is tagged `@ContinueOnFailure`, it will execute further steps.
* If not, the execution of scenario is halted.

Comments/Alternatives?

**Comment by apoorvam** [comment]
### Solution 2: 

* The annotation `ContinueOnFailure` would take parameters which gives the list of exception classes on which it would continue on failure. If its any other type of error, it breaks the execution. 
```java
@ContinueOnFailure({AssertionError, CustomError})
@Step("hello")
public void sayHello() {
    // code here
}
```
* User would still be able to do a parameterless `@ContinueOnFailure`, which would continue on any type of failure/exception.

**Comment by apoorvam** [comment]
With [this](https://github.com/getgauge/gauge-java/commit/c1b4230e18e295ec143d78ad844fb1d007785442) commit, user can specify the list of error classes as a parameter to `@ContinueOnFailure` on which it would continue on failure. 
Examples:

```java
@ContinueOnFailure({AssertionError.class, CustomError.class})
@Step("hello")
public void sayHello() {
    // code here
}
```

```java
@ContinueOnFailure(AssertionError.class)
@Step("hello")
public void sayHello() {
    // code here
}
```

In case no parameters are passed to `@ContinueOnFailure` annotation, by default, it continues with execution of further steps on any type of error. More specific, it continues for any type of error class which is implemented by base exception class in Java, i.e, `Throwable.class`.
So, parameterless `@ContinueOnFailure` annotation would still work. 

```java
@ContinueOnFailure
@Step("hello")
public void sayHello() {
    // code here
}
```

With this enhancement, user will be able to control on what type of errors the execution should continue, instead of just continuing on every type of error. For instance, on a `RuntimeException` its ideally not expected to continue further. Whereas if its an assertion error, it would be fine to continue execution.

**Comment by apoorvam** [comment]
Fix should be available in nightly >= 12.09.2016. Added functional test under `gauge-tests` [here](https://github.com/getgauge/gauge-tests/blob/1b49a0588e302e4456ac5b33a4a18555885fdb78/specs/advancedTopics/continueOnFailure/continue_on_failure_with_params.spec).

**Comment by sguptatw** [comment]
Test cases
- [x] Works only along with annotation marked as step
- [x] When no parameters are passed ContinueOnFailure continues for any kind of failure
- [x] When a whitelist error(s) class is passed as a parameter(s), it fails for any other types of errors.
- [x] When a whitelist error(s) class is passed as a parameter(s), it passes for any subtypes of the mentioned errors.
